File: exo4_1.py
import os
import sys

file_path = "test.txt"
t1_r, t1_w = os.pipe()
t2_r, t2_w = os.pipe()

fils = os.fork()

if fils == 0: # --- LE FILS ---
    os.close(t1_r)
    os.close(t2_w)
    
    word = input("Entrez le mot à chercher : ")
    os.write(t1_w, word.encode())
    res = os.read(t2_r, 1).decode()
    print(f"[Fils] Résultat reçu du père : {'Trouvé' if res == '1' else 'Non trouvé'}")
    
    os.close(t1_w)
    os.close(t2_r)

else: # --- LE PERE ---
    os.close(t1_w)
    os.close(t2_r)
    word_to_search = os.read(t1_r, 1024).decode()
    command = f"grep -q '{word_to_search}' {file_path}"
    status = os.system(command)
    
    if status == 0:
        os.write(t2_w, b"1")
    else:
        os.write(t2_w, b"0")
        
    os.close(t1_r)
    os.close(t2_w)
    os.wait() 

File: exo4_1.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/wait.h>

int main ()
{
	int p[2];
	pid_t pid;
	char buffer[1024];
	char *message = "Message du pere au fils via le tube";

	if (pipe(p) == -1) {
		printf ("erreur pipe\n");
		exit(EXIT_FAILURE);
	}

	pid = fork();
	if (pid == -1) {
		printf ("erreur fork\n");
		exit(EXIT_FAILURE);
	}

	if (pid == 0) {
		close(p[1]);
		int n = read(p[0], buffer, sizeof(buffer));
		if (n > 0) {
			buffer[n] = '\0';
			printf ("Fils (PID%d) a lu: %s\n", getpid(), buffer);
		}
		close(p[0]);
		exit(EXIT_SUCCESS);
	}
	else {
		close(p[0]);
		write(p[1], message, strlen(message));
		close(p[1]);
		wait(NULL);
	}
	return 0;
}
